### PR TITLE
Fix surrogate not using original request

### PR DESCRIPTION
**Pull request by Toflar**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Warning: This might need some close attention. It took me hours to wrap my head around that issue :-)

So the problem is that `HttpCache::forward()` (or essentially any part in your application) can modify the `$request` that is passed to the `HttpCache::handle()` and the surrogate can never access the original request using `HttpCache::getRequest()` anymore.

Example:

* Main request (GET `/foobar`)
* It's not in the cache, so `HttpCache::forward()` modifies `REMOTE_ADDR` to `127.0.0.1` and adds the `X-Forwarded-For` header.
* The request is sent to the application and any e.g. `kernel.request` listener might modify the `$request` further.
* Now the `/foobar` route returns `text/html` that contains some `<esi src="=/fragment_path"` tag.
* `HttpCache` has an instance of `SurrogateInterface` so (in our case `Esi`) will be asked to `process()` and then later on `handle()` the `/fragment_path` request. For that, `Esi` (or in fact `AbstractSurrogate` uses the following line to create a subrequest and pass it on to the application again:

```php
$subRequest = Request::create($uri, Request::METHOD_GET, array(), $cache->getRequest()->cookies->all(), array(), $cache->getRequest()->server->all());
```

What you can see here, is that it uses `$cache->getRequest()`. And here follows the problem:
We did not duplicate (clone) the original request so essentially `$cache->getRequest()` is a reference to the current request that `HttpKernel::forward()` modified and probably any other part of the application did so too. So for example the original `REMOTE_ADDR` (client IP) got lost.

What we should do instead is duplicate the original request so the surrogates can actually behave like a real reverse proxy such as Varnish would by keeping all the original request attributes.

### CODE REVIEW

**Comment by Toflar** [comment]
AppVeyor failures look unrelated and the license issue with fabbot.io is something I think I shouldn't fix?

**Review by nicolas-grekas — APPROVED** [review]
(for 2.8!)

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php [hunk 1]
```diff
@@ -165,7 +165,11 @@ public function handle(Request $request, $type = HttpKernelInterface::MASTER_REQ
         // FIXME: catch exceptions and implement a 500 error page here? -> in Varnish, there is a built-in error page mechanism
         if (HttpKernelInterface::MASTER_REQUEST === $type) {
             $this->traces = array();
-            $this->request = $request;
+            // Keep a duplicate of the original request for surrogates so they can access it.
+            // We must duplicate here to get a separate instance because the application will modify the request during
+            // the application flow (we know it always does because we do ourselves by setting REMOTE_ADDR to 127.0.0.1
+            // and adding the X-Forwarded-For header, see HttpCache::forward()).
+            $this->request = $request->duplicate();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: why duplicate and not clone? (clone is used elsewhere in the class)

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: also: should we then use the duplicated/cloned request instead of $request directly?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: let's use clone then :)

Comment (Toflar) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: Done in dcb3833b90b8f91634223a542d0c1d91a9c04800.

**Comment by nicolas-grekas** [comment]
Thank you @Toflar.

**Pull request merged**